src/palace/manager/opds/util.py

```diff
@@ -0,0 +1,91 @@
+from __future__ import annotations
+
+from collections.abc import Sequence
+from typing import Annotated, Any, TypeVar
+
+from pydantic import Discriminator, Tag
+
+from palace.manager.opds.base import BaseOpdsModel
+
+T = TypeVar("T")
+
+
+def obj_or_tuple_to_tuple(value: T | list[T] | tuple[T, ...] | None) -> Sequence[T]:
+    """
+    Convert object or sequence (list or tuple) of objects to a tuple of objects.
+    """
+    if value is None:
+        return tuple()
+    if isinstance(value, (list, tuple)):
+        return tuple(value)
+    return (value,)
+
+
+# See these links for more information about Pydantic type discriminators:
+# https://docs.pydantic.dev/latest/concepts/types/#generics
+# https://docs.pydantic.dev/latest/concepts/unions/#discriminated-unions-with-callable-discriminator
+def _discriminate_opds_type_union(value: Any) -> str | None:
+    if isinstance(value, (str, bytes, bytearray)):
+        return "string"
+    if isinstance(value, (dict, BaseOpdsModel)):
+        return "OpdsObject"
+    if isinstance(value, (list, tuple)):
+        return "list"
+    return None
+
+
+# Some type vars used in the TypeAlias definitions below.
+# Once https://github.com/pydantic/pydantic/issues/9418 is resolved
+# StrT could use a default as well as a bound, so we would only need
+# to specify it in cases where we are using a different string type.
+StrT = TypeVar("StrT", bound=str)
```

Comment:
It looks like [the fix for this](https://github.com/pydantic/pydantic/pull/10789) has been merged, so soon, hopefully.